### PR TITLE
CSP-389-auditing-DeleteStandard-inner

### PR DESCRIPTION
**Pull request by raviudari01**

### CODE REVIEW

File: src/SFA.DAS.Roatp.Data/Repositories/AuditWriteRepository.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                UpdatedState = JsonConvert.SerializeObject(entityUpdatedState, jsonSerializerSettings),
+                AuditDate = DateTime.Now
+            };
+            _roatpDataContext.Audits.Add(audit);
+        }
+
+        public void AddAudit(List<T> entityInitialState, List<T> entityUpdatedState, string entityId, string userId, string userDisplayName, string userAction )
+        {
+            var jsonSerializerSettings = new JsonSerializerSettings
+            {
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/AuditWriteRepository.cs:44]: Duplicate code.

File: src/SFA.DAS.Roatp.Data/Repositories/ProviderCoursesWriteRepository.cs [hunk 2]
```diff
@@ -43,16 +47,31 @@ public async Task<ProviderCourse> CreateProviderCourse(ProviderCourse providerCo
...
+            await using var transaction = await _roatpDataContext.Database.BeginTransactionAsync();
+            try
+            {
+                var providerCourse = await _roatpDataContext.ProviderCourses
                 .Where(l => l.LarsCode == larscode && l.Provider.Ukprn == ukprn)
                 .Include(l => l.Locations).Include(l => l.Versions)
                 .SingleAsync();
 
-            _roatpDataContext.Remove(providerCourse);
+                AddAudit(providerCourse, null, providerCourse.Id.ToString(), userId, userDisplayName, userAction);
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ProviderCoursesWriteRepository.cs:60]: This literally can become a constructor function in Audit repository. You can avoid the base repository (I know I suggested, but you implemented ;-) ). and add the audit record here itself to the context. 
Also if you go down this path, the constructor just needs to take object type for the first two arguments and serialise the objects (one or many does not matter), so you can remove the code duplication.

File: src/SFA.DAS.Roatp.Data/Repositories/AuditWriteRepository.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            var audit = new Audit
+            {
+                CorrelationId = Guid.Parse(Activity.Current.RootId),
+                EntityType = typeof(T).Name,
+                UserAction = userAction,
+                UserId = userId,
+                UserDisplayName = userDisplayName,
+                EntityId = entityId,
+                InitialState = JsonConvert.SerializeObject(entityInitialState, jsonSerializerSettings),
+                UpdatedState = JsonConvert.SerializeObject(entityUpdatedState, jsonSerializerSettings),
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/AuditWriteRepository.cs:35]: Why not use in-built System.Text.Json namespace?

File: src/SFA.DAS.Roatp.Data/SFA.DAS.Roatp.Data.csproj [hunk 4]
```diff
@@ -10,6 +10,7 @@
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.8" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.8" />
     <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="6.0.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/SFA.DAS.Roatp.Data.csproj:13]: Use framework library System.Text.Json

File: src/SFA.DAS.Roatp.Data/Repositories/AuditWriteRepository.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+        public void AddAudit(T entityInitialState, T entityUpdatedState, string entityId, string userId, string userDisplayName, string userAction)
+        {
+            var jsonSerializerSettings = new JsonSerializerSettings
+            {
+                PreserveReferencesHandling = PreserveReferencesHandling.Objects
+            };
+            var audit = new Audit
+            {
+                CorrelationId = Guid.Parse(Activity.Current.RootId),
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/AuditWriteRepository.cs:28]: Would two different calls on two different threads still share the same activity id? You can check by retrieving this in the controller/handler and repository to see if all the ids match

File: src/SFA.DAS.Roatp.Database/Tables/Audit.sql [hunk 6]
```diff
@@ -0,0 +1,18 @@
+﻿CREATE TABLE [dbo].[Audit]
+(
+	[Id] BIGINT NOT NULL IDENTITY(1,1),
+	[CorrelationId] [uniqueidentifier] NULL,
+	[EntityType] NVARCHAR(256) NOT NULL,
+	[EntityId] NVARCHAR(256),
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Database/Tables/Audit.sql:6]: Specify NULL/NOT NULL on all the fields to keep it consistent

**Review by MarkFCain — APPROVED** [review]

File: src/SFA.DAS.Roatp.Data/Repositories/ProviderLocationsBulkRepository.cs [hunk 7]
```diff
@@ -25,15 +29,29 @@ public async Task BulkInsert(IEnumerable<ProviderLocation> providerLocations)
...
+        public async Task BulkDelete(IEnumerable<int> providerLocationIds, string userId, string userDisplayName, int ukprn, string userAction)
         {
-            var providerLocations = await _roatpDataContext.ProviderLocations
+            await using var transaction = await _roatpDataContext.Database.BeginTransactionAsync();
+            try
+            {
+                var providerLocations = await _roatpDataContext.ProviderLocations
                 .Where(l => providerLocationIds.Contains(l.Id))
-                .ToListAsync();
+            .ToListAsync();
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ProviderLocationsBulkRepository.cs:39]: Alignment is messed up

**Review by DevenKShah — COMMENTED** [review]
Few suggestions

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**